packages/muelu/src/Misc/MueLu_AggregateQualityEstimateFactory_def.hpp

```diff
@@ -320,31 +339,50 @@ namespace MueLu {
       MT* vr = NULL;
 
       const char compute_flag ='N';
-      myLapack.GGES(compute_flag,compute_flag,compute_flag,ptr2func,aggSize,
-                    topMatrix.values(),aggSize,bottomMatrix.values(),aggSize,&sdim,
-                    alpha_real.values(),alpha_imag.values(),beta.values(),vl,aggSize,
-                    vr,aggSize,workArray.values(),workArray.length(),bwork,
-                    &info);
-
-      TEUCHOS_ASSERT(info == LO_ZERO);
-
-      MT maxEigenVal = MT_ZERO;
-
-      for (int i=LO_ZERO;i<aggSize;++i) {
-        // NOTE: In theory, the eigenvalues should be nearly real
-        //TEUCHOS_ASSERT(fabs(alpha_imag[i]) <= 1e-8*fabs(alpha_real[i])); // Eigenvalues should be nearly real
-        maxEigenVal = std::max(maxEigenVal, alpha_real[i]/beta[i]);
-
+      if(algo == ALG_FORWARD) {
+        // Forward: Solve the generalized eigenvalue problem as is
+        myLapack.GGES(compute_flag,compute_flag,compute_flag,ptr2func,aggSize,
+                      topMatrix.values(),aggSize,bottomMatrix.values(),aggSize,&sdim,
+                      alpha_real.values(),alpha_imag.values(),beta.values(),vl,aggSize,
+                      vr,aggSize,workArray.values(),workArray.length(),bwork,
+                      &info);
+        TEUCHOS_ASSERT(info == LO_ZERO);
+
+        MT maxEigenVal = MT_ZERO;
+        for (int i=LO_ZERO;i<aggSize;++i) {
+          // NOTE: In theory, the eigenvalues should be nearly real
+          //TEUCHOS_ASSERT(fabs(alpha_imag[i]) <= 1e-8*fabs(alpha_real[i])); // Eigenvalues should be nearly real
+          maxEigenVal = std::max(maxEigenVal, alpha_real[i]/beta[i]);          
+        }
+        
+        (agg_qualities->getDataNonConst(0))[aggId] = (MT_ONE+MT_ONE)*maxEigenVal;
       }
-
-      (agg_qualities->getDataNonConst(0))[aggId] = (MT_ONE+MT_ONE)*maxEigenVal;
-
-    }
-
-
+      else {
+        // Reverse: Swap the top and bottom matrices for the generalized eigenvalue problem
+        // This is trickier, since we need to grab the smallest non-zero eigenvalue and invert it.
+        myLapack.GGES(compute_flag,compute_flag,compute_flag,ptr2func,aggSize,
+                      bottomMatrix.values(),aggSize,topMatrix.values(),aggSize,&sdim,
+                      alpha_real.values(),alpha_imag.values(),beta.values(),vl,aggSize,
+                      vr,aggSize,workArray.values(),workArray.length(),bwork,
+                      &info);
+
+        TEUCHOS_ASSERT(info == LO_ZERO);
+        
+        MT minEigenVal = MT_ZERO;
+       
+        for (int i=LO_ZERO;i<aggSize;++i) {
+          MT ev = alpha_real[i] / beta[i];
```

Comment:
Do we worry about beta[i] ever being zero?

Reply:
We never did before...